Warn experimentalist when Ramsey freq change > artificial detuning

stephlazar opened the pull request:
Adds feature to warn the experimentalist if the qubit frequency change from a Ramsey experiment is larger than the chosen artificial detuning:

- some text in the Ramsey image text box will be highlighted in red
- an image containing a warning will be copied to the data folder
- can be tested with: a = tda.RamseyAnalysis(t_start='20210603_112017', options_dict={'artificial_detuning': 10e3})

Commit:
BaseDataAnalysis: adds _raise_warning_image + adds analysis_v2/WARNING.png

Commit:
analysis/tools/plotting.py: adds rainbow_text to allow multi-coloured strings

Commit:
Ramsey: raise warning when change in frequency larger than artificial_detuning

Commit:
Adds docstrings to BaseDataAnalysis._raise_warning_image and RamseyAnalysis.prepare_plots

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -5305,16 +5310,29 @@ def prepare_plots(self):
+                        raised_warning = True
+                    textstr = textstr.split('\n')
+                    sc = [s for s in textstr if 'Delta f' in s][0]
+                    idx = textstr.index(sc)
```
Isn't the code easier as follows?
```
idx = [i for i, s in enumerate(textstr) if 'Delta f' in s][0]
```

stephlazar replied:
Yes, thanks

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -5305,16 +5310,29 @@ def prepare_plots(self):
+                    idx = textstr.index(sc)
+                    color = ['black']*len(textstr)
+                    color[idx] = 'red'
+                    color[-1] = 'red'
```
Please add a comment that -1 is the line "artificial detuning" (or would it make sense to search for the index like you do for the other line?)

stephlazar replied:
Yes, it makes sense to find index here too.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -5305,16 +5310,29 @@ def prepare_plots(self):
+
+                color = 'k'
+                if np.abs(delta_f) > np.abs(art_det):
+                    # we don't want this; raise a warning and highlight in red
```
maybe you could replace "we don't want this" by a physics explanation

stephlazar replied:
Indeed.

chellings commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -5305,16 +5310,29 @@ def prepare_plots(self):
+                    # the Delta f and artificial detuning row in textstr
+                    if not raised_warning:
+                        self._raise_warning_image()
+                        raised_warning = True
```
maybe self._raise_warning_image could automatically keep track of whether a warning has already been raised (e.g., by setting a self._warning_image_raised private property). Having the check there would make the code easier in all functions that call _raise_warning_image

stephlazar replied:
Makes sense 👍

chellings commented on pycqed/analysis/tools/plotting.py:
```diff
@@ -370,3 +371,38 @@ def autolabel_barplot(ax, rects, rotation=90):
                 ha='center', va='bottom', rotation=rotation)
+
+
+def rainbow_text(x, y, strings, colors, ax, orientation='horizontal', **kwargs):
```
please add a comment with the source when copying a function from the web

stephlazar replied:
👍

chellings commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -241,6 +243,25 @@ def run_analysis(self):
                 log.error("Unhandled error during analysis!")
                 log.error(traceback.format_exc())
 
+    def _raise_warning_image(self, image_path=None):
```
As discussed orally: we might want to move the core functionality to a function in utilities. This function would require a destination folder. We could then have a wrapper here, which calls the utility function passing the folder of the current analysis. Then, in the future, other modules that know a timestamp folder (e.g., MC) could make use of the same utility function.

stephlazar replied:
Makes sense, I moved it now to utilities.py.

chellings commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -241,6 +243,25 @@ def run_analysis(self):
+            image_path = os.path.abspath(
+                sys.modules[self.__class__.__module__].__file__)
+            image_path = os.path.split(image_path)[0]
+            image_path = os.path.abspath(os.path.join(image_path, 'WARNING.png'))
```
While we might find the text on the waning image funny now, wouldn't a simpler warning sign (e.g., just the red warning text, or a yellow sign with exclamation mark) be a more sustainable solution?

Moreover, as discussed orally, it would make sense to log a warning message into a text file to inform the user what triggered the warning sign. (Note that you would need to keep appending to the text file if this function is called multiple times; even if you don't want to copy the warning image a second time.)

stephlazar replied:
I added this functionality now.

chellings reviewed (changes requested):
Thanks, @stephlazar, for implementing this. As already discussed orally, I would have some suggestions. See my comments.

Commit:
Merge branch 'qudev_master' into enh/ramsey_art_det_warning

Commit:
Merge branch 'qudev_master' into enh/ramsey_art_det_warning

Commit:
RamseyAnalysis: improves docstring and code readability when raising warning in the case qubit detuning > artificial detuning.

Commit:
rainbow_text: adds comment with source

Commit:
BaseDataAnalysis,RamseyAnalysis: adds private property _warning_image_raised and checks it in _raise_warning_image + removes this flag and check from RamseyAnalysis

Commit:
general.py: adds raise_warning_image, BaseDataAnalysis: _raise_warning_image calls this function now in general.py

Commit:
general.py, BaseDataAnalysis, RamseyAnalysis: adds options to write a text file with a warning message

- general.py: adds write_warning_message_to_text_file
- BaseDataAnalysis: replaces _raise_warning_image by _raise_warning which now calls both raise_warning_image and write_warning_message_to_text_file
- adds warning message in call to _raise_warning in RamseyAnalysis.
- replaces the old WARNING.png with a simpler one

chellings commented on pycqed/utilities/general.py:
```diff
@@ -800,3 +800,65 @@ def find_symmetry_index(data):
+        the module folder.
+    :return:
+    """
+    if 'WARNING.png' not in os.listdir(destination_path):
```
Since you have the check here, you do not need the check with _warning_image_raised in ba

stephlazar replied:
Done

chellings commented on pycqed/utilities/general.py:
```diff
@@ -800,3 +800,65 @@ def find_symmetry_index(data):
+    """
+    if filename is None:
+        filename = 'warning_message'
+    filename += '.txt'
```
I would do this only if the filename does not have an extension yet

stephlazar replied:
Changed now

chellings commented on pycqed/utilities/general.py:
```diff
@@ -800,3 +800,65 @@ def find_symmetry_index(data):
+        filename = 'warning_message'
+    filename += '.txt'
+
+    # Adding this will ensure the if a future message is appended to the file,
```
"ensure the if" -- > "ensure that if"

stephlazar replied:
👍 thanks

chellings commented on pycqed/utilities/general.py:
```diff
@@ -800,3 +800,65 @@ def find_symmetry_index(data):
+        file_message = file.read()
+        file.close()
+        if message in file_message:
+            # Do not add message if already exists.
```
Why not? If a warning happens multiple times, I would expect to see it multiple times in the file (same behavior as the normal logging system)

stephlazar replied:
My thought here was to prevent that the same message is appended each time we reanalyze the file offline. Do you want it to do this?

stephlazar replied:
Could you have another look at this before I merge? I now made some bigger changes.

This is what the warning file looks like now:
![image](https://user-images.githubusercontent.com/14844492/127144986-f34313f3-68e9-4815-8f91-abfdf0d385af.png)

chellings reviewed (approved):
Thanks for the updated version. I have added some final comments, and I would approve to merge after those are addressed.

Commit:
Merge branch 'qudev_master' into enh/ramsey_art_det_warning

Commit:
BaseDataAnalysis: removes _warning_image_raised attribute

Commit:
general.py/write_warning_message_to_text_file: add filename extension only if not there + docstring improvements

Commit:
BaseDataAnalysis._raise_warning: updates docstring

Commit:
BaseDataAnalysis._raise_warning: save warning image + text file only when delegate_plotting is False

Commit:
general.py/write_warning_message_to_text_file: prepend timestamp to message

Commit:
BaseDataAnalysis, RamseyAnalysis: changes related to raising warnings

- creates the attributes _warning_message and _raise_warning_image
- children append message to self._warning_message and set the flag self._raise_warning_image = True
- adds call to self._raise_warning at the end of self.run_analysis
- makes warning_message and warning_textfile_name settable in self._raise_warning by taking them from options_dict

Commit:
BaseDataAnalysis._raise_warning: fixes small bug

chellings commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -234,13 +242,50 @@ def run_analysis(self):
+        warning image to the folder corresponding to the last timestamp in
+        self.timestamps.
+
+        If warning_message is not None, calls write_warning_message_to_text_file
```
docstring is outdated

chellings commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -234,13 +242,50 @@ def run_analysis(self):
+        """
+        if not self.check_plotting_delegation():
+            # else, the warning image and text file will be generated twice
+            # when the AnalysisDaemon is active
```
Just some general comments (do not necessarily need to be addressed in this PR):
- Explaining an if statement with a comment that starts with the word "else" might cause confusion.
- I wonder what the best place for such an explanation is: in the docstring or here.
- I would prefer the following code in such an instance to avoid an additional level of indentation. It does not really matter in this special case since there are no deep indentation levels in this function, so it is, as I said, just a general comment
```
if self.check_plotting_delegation():
    return
```

chellings reviewed (approved):
Looks good to me. Just some minor comments, which you could consider before merging.

Commit:
BaseDataAnalysis._raise_warning: improves docstring + minor cleanup
